lib/semantics/mod-file.cc

```diff
@@ -590,84 +608,116 @@ std::ostream &PutLower(std::ostream &os, const std::string &str) {
-  std::fstream stream;
-  auto header{GetHeader(contents)};
-  auto size{GetFileSize(path)};
-  if (size == header.size() + 1 + contents.size()) {
-    // file exists and has the right size, check the contents
-    stream.open(path, std::ios::in | std::ios::out);
-    if (FileContentsMatch(stream, header, contents)) {
-      return true;
-    }
-    stream.seekp(0);
-  } else {
-    stream.open(path, std::ios::out);
+struct Temp {
+  Temp() = delete;
+  ~Temp() {
+    close(fd);
+    unlink(path.c_str());
+  }
+  int fd;
+  std::string path;
+};
+
+// Create a temp file in the same directory and with the same suffix as path.
+// Return an open file descriptor and its path.
+static Temp MkTemp(const std::string &path) {
+  auto length{path.length()};
+  auto dot{path.find_last_of("./")};
+  std::string suffix{dot < length && path[dot] == '.' ? path.substr(dot) : ""};
+  CHECK(length > suffix.length() &&
+      path.substr(length - suffix.length()) == suffix);
+  auto tempPath{path.substr(0, length - suffix.length()) + "XXXXXX" + suffix};
+  int fd{mkstemps(&tempPath[0], suffix.length())};
+  auto mask{umask(0777)};
+  umask(mask);
+  chmod(tempPath.c_str(), 0666 & ~mask);  // temp is created with mode 0600
+  return Temp{fd, tempPath};
+}
+
+// Write the module file at path, prepending header. If an error occurs,
+// return an error message, otherwise nullptr.
+static int WriteFile(const std::string &path, const std::string &contents) {
+  auto header{std::string{ModHeader::bom} + ModHeader::magic +
+      CheckSum(contents) + ModHeader::terminator};
+  if (FileContentsMatch(path, header, contents)) {
+    return 0;
+  }
+  Temp temp{MkTemp(path)};
+  if (temp.fd < 0) {
+    return errno;
+  }
+  if (write(temp.fd, header.c_str(), header.size()) == -1 ||
+      write(temp.fd, contents.c_str(), contents.size()) == -1) {
+    return errno;
   }
-  stream << header << '\n' << contents;
-  stream.close();
-  return !stream.fail();
+  if (std::rename(temp.path.c_str(), path.c_str()) == -1) {
+    return errno;
+  }
+  return 0;
 }
 
 // Return true if the stream matches what we would write for the mod file.
-static bool FileContentsMatch(std::fstream &stream, const std::string &header,
-    const std::string &contents) {
-  char c;
-  for (std::size_t i{0}; i < header.size(); ++i) {
-    if (!stream.get(c) || c != header[i]) {
-      return false;
-    }
+static bool FileContentsMatch(const std::string &path,
+    const std::string &header, const std::string &contents) {
+  std::size_t hsize{header.size()};
+  std::size_t csize{contents.size()};
+  if (GetFileSize(path) != hsize + csize) {
+    return false;
   }
-  if (!stream.get(c) || c != '\n') {
+  int fd{open(path.c_str(), O_RDONLY)};
+  if (fd < 0) {
     return false;
   }
-  for (std::size_t i{0}; i < contents.size(); ++i) {
-    if (!stream.get(c) || c != contents[i]) {
+  constexpr std::size_t bufSize{4096};
+  std::string buffer(bufSize, '\0');
+  if (read(fd, &buffer[0], hsize) == -1 ||
```

Comment:
Beware of short reads.  It's better to compare the result of `read()` with the expected byte count.